src/providers/mod.rs

```diff
@@ -0,0 +1,124 @@
+use failure::Error;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct LaunchDescriptor<M: MachineSetup> {
+    pub region: M::Region,
+    pub log: slog::Logger,
+    pub max_wait: Option<std::time::Duration>,
+    pub machines: Vec<(String, M)>,
+}
+
+/// This is used to group machines into connections
+/// to cloud providers. e.g., for AWS we need a separate
+/// connection to each region.
+pub trait MachineSetup {
+    type Region: Eq + std::hash::Hash + Clone + ToString;
```

Comment:
I'm surprised the linter doesn't complain about lack of docs here? Is the module perhaps not `pub`?